company.el

```diff
@@ -1558,54 +1537,66 @@ Only affects `company-text-icons-margin'."
   :type 'string)
 
 (defcustom company-text-icons-add-background nil
-  "When non-nil, add special background to the characters."
+  "When non-nil, generate a background color for text icons when none is given.
+See `company-text-icons-mapping'."
   :type 'boolean)
 
 (defun company-text-icons-margin (candidate selected)
   "Margin function which returns unicode icons."
   (when-let ((candidate candidate)
              (kind (company-call-backend 'kind candidate))
-             (icon (or (alist-get kind company-text-icons-mapping)
-                       (alist-get t company-text-icons-mapping)))
-             (face (or (assoc-default kind
-                                      company-text-kind-face-mapping)
-                       (assoc-default t company-text-kind-face-mapping))))
-    (propertize
-     (format company-text-icons-format icon)
-     'face
-     `(,@company-text-face-extra-attributes
-       ,@(when company-text-icons-add-background
-           (list :background
-                 (company-text-icons--background face selected)))
-       :foreground ,(face-attribute face :foreground)))))
+             (conf (or (alist-get kind company-text-icons-mapping)
+                       (alist-get t company-text-icons-mapping))))
+    (cl-destructuring-bind (icon &optional fg bg) conf
+      (propertize
+       (format company-text-icons-format icon)
+       'face
+       (company-text-icons--face fg bg selected)))))
 
 (declare-function color-rgb-to-hex "color")
 (declare-function color-gradient "color")
 
-(defun company-text-icons--background (face selected)
-  (apply #'color-rgb-to-hex
-         (nth 0 (color-gradient
-                 (color-name-to-rgb
-                  (face-attribute
-                   (if selected
-                       'company-tooltip-selection
-                     'company-tooltip)
-                   :background))
-                 (color-name-to-rgb
-                  (face-attribute face :foreground))
-                 10))))
+(defun company-text-icons--face (fg bg selected)
+  ;; Narrow to specific bg used for current candidate when a CONS cell.
+  (when (consp bg)
```

Comment:
Very nice.

There is one case that doesn't seem to be handled, though: what happens if `bg` is not a cons and `selected` is t?

We could still try to auto-generate the color for it based on the colors of `bg` and `company-tooltip-selection`, using a similar interpolation, though perhaps with a different number of steps. The exact number needs some fiddling and would probably never be ideal, but that still seems better than not extending the selection rectangle to the left border for those entries. WDYT?

Reply:
> But there is one case that doesn't seem to be handled: what happens if bg is not a cons and selected is t?

Then bg would (ideally) be a string, face symbol or nil. If it's a string then `:background` is set to it. If it's a face then we use that faces :background property or if that doesn't exist then we use no background. If it's nil then we generate a background using the fg-color and the tooltip bg. 

I'm not sure what condition you mean in this comment.

EDIT: I got it by the time I reach the end of second paragraph below I think.

> but that still seems better than not extending the selection rectangle

Do you mean not having a separate face for the icon in the margin when we're in `company-tooltip-selection`? Ideally I'd suggest the user supply a separate bg color (using the cons cell) if they want it. If they don't set any background thten we generate separate ones for when the candidates selected and when it isn't. If they supply only one symbol for BG my guess was that they didn't want separate colors for selected and !selected backgrounds.

Are you suggesting when `selected` is `t` and BG isn't a cons cell that we should generate another background to indicate that this is the current selection? Honestly seems kinda like an overreach to me. I'd expect if someone went to the effort of specifying their own BG colors in the config, they could go the extra mile and also specify BG for when the current candidate is selected.

Reply:
> If they supply only one symbol for BG my guess was that they didn't want separate colors for selected and !selected backgrounds.

It might be a plausible guess, but I'm not sure there is any significant number of users who would prefer such behavior. We could guard this auto-generation behind `(when company-text-icons-add-background ...)`  as well.

> Are you suggesting when selected is t and BG isn't a cons cell that we should generate another background to indicate that this is the current selection? Honestly seems kinda like an overreach to me. I'd expect if someone went to the effort of specifying their own BG colors in the config, they could go the extra mile and also specify BG for when the current candidate is selected.

I was thinking some users might want to emulate what Atom does (I showed the screenshot in the other thread): pick ~5 simple, pleasant background colors, associate each of the kinds with one of them, and let our code do the remaining blending. That would be a low-effort configuration that could, however, look nice if we do this extra step.

Further, the `company-tooltip-selection` colors come from the theme. I'm guessing not everybody will want to change their `company-text-icons-mapping` customization as they switch between themes. Whereas, if they only need to specify the base color, the result should look good enough across a whole number of dark themes (for example). They'll need to choose different base colors after switching to a light theme, though.

Reply:
@dgutov 

Okay, I understand the motivations. So just referring to the implementation how do we approach this. If there *is* a background (eg: string or face, not cons) do we blend from the background instead of the foreground, or would we prefer blending from the foreground? Just curious what precedence we're going for here.

Reply:
From the background.

Using some different number of "steps" rather than 10, probably. Maybe 2-4? That needs some testing.

Reply:
> Using some different number of "steps" rather than 10, probably. Maybe 2-4? That needs some testing.

I was hoping we could add a user option to configure the steps. `company-text-icons-bg-gradient-step`, I think I might've changed the steps from version you had before by accident cause I'm finding the contrast a little weaker.

Reply:
Alright, done. It's kind of ugly IMO but it works.

Reply:
Thank you, looks good. I'll try it out a bit later and merge.

> I was hoping we could add a user option to configure the steps. company-text-icons-bg-gradient-step

No objection from me, though now we have two values of STEPS to possibly configure. I guess we could allow `company-text-icons-add-background` to take a list value as well (with numbers). The first number would be the override for 10, the second one will override 2.

Reply:
Well for now using 10 and 2 looks well enough. If there comes a demand down the line for customization we can add the options then but for now it all looks fine to me.